src/lib/Inference.hs

```diff
@@ -1639,6 +1647,30 @@ withSuperclassBindersRec (ty:rest) cont = do
     withSuperclassBindersRec rest' cont
   return $ Abs (Nest b bs) e
 
+withNestedUBindersArrow
+  :: forall i i' o e. (EmitsInf o, HasNamesE e, SubstE AtomSubstVal e, SinkableE e)
+  => Nest (UAnnBinderArrow AtomNameC) i i'
+  -> (forall o'. (EmitsInf o', Ext o o') => [AtomName o'] -> InfererM i' o' (e o'))
+  -> InfererM i o (Abs (Nest PiBinder) e o)
+withNestedUBindersArrow bs cont = case bs of
+  Empty -> Abs Empty <$> cont []
+  Nest (UAnnBinderArrow b ty arr) rest -> do
+    Abs (b':>ty') (Abs rest' body) <-
+      case arr of
+        PlainArrow -> withUBinder (UAnnBinder b ty) id inner_c
+        ClassArrow -> withUBinder (UAnnBinder b ty)
+          (LamBound . LamBinding ClassArrow) inner_c
+        ImplicitArrow -> withUBinder (UAnnBinder b ty)
+          (LamBound . LamBinding ImplicitArrow) inner_c
+        LinArrow -> withUBinder (UAnnBinder b ty) id inner_c
+    return $ Abs (Nest (PiBinder b' ty' arr) rest') body
```

Comment:
Wouldn't it be simpler to do
```hs
let binding = case arr of PlainArrow -> id; ClassArrow -> ...
Abs (b':>ty') (Abs rest' body) <- withUBinder (UAnnBinder b ty) binding do
  cont =<< ((:names) <$> sinkM name)
```

Reply:
Thanks!  Using `toBinding` instead of `id` (as @apaszke suggested offline) made the type-checker happy.